f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py

```diff
@@ -161,9 +161,81 @@ def _create(self, bigip, vs, listener, service):
         network_id = loadbalancer.get('network_id', "")
         self.driver.service_adapter.get_vlan(vs, bigip, network_id)
         super(ListenerManager, self)._create(bigip, vs, listener, service)
-
+        
     @log_helpers.log_method_call
     def delete(self, listener, service, **kwargs):
         self._search_listener(listener, service)
         payload = self.driver.service_adapter.get_virtual_name(service)
         super(ListenerManager, self).delete(listener, service, payload=payload)
+
+
+
+class PoolManager(ResourceManager):
+
+    def __init__(self, driver):
+        super(PoolManager, self).__init__(driver)
+        self._resource = "pool"
+        self.resource_helper = resource_helper.BigIPResourceHelper(
+            resource_helper.ResourceType.pool)
+        self.mutable_props = {
+            "name": "description",
+            "description": "description",
+            "lb_algorithm": "loadBalancingMode"
+        }
+
+    def _create_payload(self, pool, service):
+        for element in service['pools']:
+            if element['id'] == pool['id']:
+                service['pool'] = element
+                break
+        if not service.get('pool'):
+            raise Exception("Invalid input: pool %s "
+                            "is not in service payload %s",
+                            pool['id'], service)
+
+        return self.driver.service_adapter.get_pool(service)
+
+    def _create(self, bigip, poolpayload, pool, service):
+        """ create the pool at first"""
+    
+        super(PoolManager, self)._create(bigip, poolpayload, pool, service)
+        
+        for listener in service['listeners']:
+            if listener['default_pool_id'] == pool['id']:
+               service['listener'] =listener 
+               break
+
+        """Update the listener's default pool id if needed"""
+        if service.get('listener'):
+            LOG.debug("Find a listener %s for create pool",listener)
+            mgr = ListenerManager(self.driver)
+            old_listener = {
+            }
+            old_listener['default_pool_id'] = ''
+            old_listener['id'] = pool['id']
+            old_listener['admin_state_up'] = listener['admin_state_up']
+            mgr.update(old_listener,listener,service)
+
+
+    def _delete(self,bigip,poolpayload,pool,service):
+
+        for listener in service['listeners']:
+            if listener['default_pool_id'] == pool['id']:
+               service['listener'] =listener 
+               break
+
+        """Update the listener's default pool id if needed"""
```

Comment:
There might be several virtual servers who link this pool as the default one. We need to update all of them.